### PR TITLE
shell.nix for building monero

### PR DESCRIPTION
**Pull request by thedavidmeister**
## changes

- adds a `shell.nix` file
  - using `nix-shell` (part of nixos) with this file pulls all deps for building monero locally
  - includes `hidapi` for ledger nano s compatibility
  - updates readme

### CODE REVIEW

**Comment by thedavidmeister** [comment]
@moneromooo-monero by "stray formatting" what do you mean? the removal of trailing whitespace?

**Comment by moneromooo-monero** [comment]
Yes please. All that's not your actual nix related changes.

File: shell.nix [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    gcc
+    pkgconfig
+    boost
+    openssl
+    zeromq3
+    cppzmq
+    unbound
+    libsodium
+    # needed for ledger nano s
+    hidapi
```
Comment (TheCharlatan) [inline shell.nix]: also needs protobuf and libusb , if you want Trezor support as well (though optional, like hidapi).

**Comment by thedavidmeister** [comment]
@moneromooo-monero why not fix them? do the trailing whitespaces signify something important? was auto from the IDE

**Comment by thedavidmeister** [comment]
@moneromooo-monero well trailing whitespace is not spurious to cleanup from the perspective of someone who works with an IDE that cleans them up on save, means i have to put the spaces back in manually

truly minimising changes would be catching the bad formatting ahead of time with a preventative measure, like an automated test, not asking downstream developers to fiddle with it when they want to work with the files

is there something we can do at the testing layer to clean up the readme in a sustainable way before working on this branch further?

**Comment by selsta** [comment]
@thedavidmeister Please read the https://github.com/monero-project/monero/blob/master/CONTRIBUTING.md this is all explained.

**Comment by TheCharlatan** [comment]
@thedavidmeister you don't need to fiddle with anything, just add the actual hunk you are changing with `git add -p`. I agree though that a common liniting tool would be useful, that we just run once over everything.

**Comment by thedavidmeister** [comment]
it says no _random_ whitespace changes.

these changes aren't random, they're perfectly predictable. a huge number of people will have an IDE configured to nuke trailing whitespace, it is the default in atom and probably others.

if sane OOTB settings for one of the most popular text editors is considered a blocker for PRs then it seems like prevention would be better than some bandaid git command that everyone can run over and over forever?

https://github.com/remarkjs/remark-lint popped up in a quick google, i bet there are other options :)

**Comment by moneromooo-monero** [comment]
Then at least hide this in a separate patch.

**Comment by thedavidmeister** [comment]
@fuwa0529 ooh yeah that appears to work

works much better for me since ledger support was added to nixpkgs recently ;)

so i wonder if this is just a matter of updating the readme then?

**Comment by fuwa0529** [comment]
That's great, and I learned this trick by looking at this PR, so thanks to you. By the way, `$NIXPKGS` can be replaced with `'<nixpkgs>'`, including the quotes, I think that will use your current build expressions.

**Comment by thedavidmeister** [comment]
@fuwa0529 that syntax works much better for me, for some reason `$NIXPKGS` is empty on my machine

so `nix-build '<nixpkgs>' --run-env -A monero` is probably the way to go then

**Comment by moneromooo-monero** [comment]
Can you rebase now please, and address the comment by TheCharlatan. Then it can go in.

**Comment by fuwa0529** [comment]
I think this can be closed now, the readme clean up patch #5036 is already merged. And we have discovered that a more consistent way to enter a build environment is to take advantage of the expression in nixpkgs, hence no `shell.nix` is needed.

**Comment by moneromooo-monero** [comment]
OK. I'll leave this open a bit more to see if the author comments.

**Comment by thedavidmeister** [comment]
yes it can be closed

unless we want to put instructions in the readme for the above process?

**Comment by moneromooo-monero** [comment]
If you think it's helpful. I don't know that system so cannot comment if nobody else who does does.

**Comment by thedavidmeister** [comment]
@moneromooo-monero my experience so far with nixos is that some ppl are very comfortable with it and would find this advice obvious, but a lot of people are at a casual/curious level of familiarity with the system and may not realise that they can use flags with `nix-build` to drop into a `nix-shell`

happy to close this out, this discussion should at least show up in google now ;)

**Pull request closed without merging**